### PR TITLE
sql: support repartitioning of secondary indexes

### LINKED ISSUE #20874: sql: support repartitionings on secondary indexes
`ALTER INDEX t@foo PARTITION BY ...`

This already works for `ALTER TABLE`

### PR DESCRIPTION
**Pull request by danhhz**
This adds and implements the new ALTER INDEX PARTITION BY ... synxtax. A
AlterIndex statement is added to mirror AlterTable, which takes a
variable number of alterations.

Closes #20874

Release note: None

### CODE REVIEW

File: pkg/sql/alter_index.go [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+		return err
+	}
+
+	// Record this index alteration in the event log. This is an auditable log
+	// event and is recorded in the same transaction as the table descriptor
+	// update.
+	if err := MakeEventLogger(params.p.LeaseMgr()).InsertEventRecord(
+		params.ctx,
+		params.p.txn,
+		// TODO(dan): BEFORE MERGE make an EventLogAlterIndex to use here
```
Comment (danhhz) [inline pkg/sql/alter_index.go]: Intentionally leaving this TODO undone until I make sure that no-one is going to have strong objections to adding this generic CreateIndex

**Comment by benesch** [comment]
:lgtm:, but I'll note that PostgreSQL doesn't support multiple `ALTER INDEX` commands in one statement like they do for `ALTER TABLE`. Perhaps get someone else to weigh in?

---

Reviewed 12 of 12 files at r1.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/sql/alter_index.go, line 44 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20957#-L0u_IlMToPg8Tomcfji:-L0u_IlMToPg8Tomcfjj:bshdkn) ([raw file](https://github.com/cockroachdb/cockroach/blob/7797919d7db3b82bb816d12f18ce3e6596a429e8/pkg/sql/alter_index.go#L44)):*
> ```Go
> 	// different copy than the one in the tableDesc. To make it easier for the
> 	// code below, get a pointer to the index descriptor that's actually in
> 	// tableDesc.
> ```

omg that must've been fun to track down

---

*[pkg/sql/alter_index.go, line 70 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20957#-L0u_N_nrsHy36RNcne8:-L0u_N_nrsHy36RNcne9:b-g6mtjj) ([raw file](https://github.com/cockroachdb/cockroach/blob/7797919d7db3b82bb816d12f18ce3e6596a429e8/pkg/sql/alter_index.go#L70)):*
> ```Go
> 			n.indexDesc.Partitioning = partitioning
> 			// TODO(benesch): install CHECK constraint to exclude rows that do
> 			// not belong to any partition of the unique constraint.
> ```

No longer needed right?

---

*[pkg/sql/alter_index.go, line 82 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20957#-L0u_eZbyFqAjJkGWS-o:-L0u_eZbyFqAjJkGWS-p:bcps7ht) ([raw file](https://github.com/cockroachdb/cockroach/blob/7797919d7db3b82bb816d12f18ce3e6596a429e8/pkg/sql/alter_index.go#L82)):*
> ```Go
> 				// TODO(dan): Consider the case of a table that is repartitioned
> 				// when it has not finished the schema change for adding the
> 				// partitioning CHECK constraint from the last partitioning.
> ```

Can drop this, right?

---

*[pkg/sql/alter_index.go, line 94 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20957#-L0u_SvwxkjFUNtAu-S2:-L0u_SvwxkjFUNtAu-S3:b7ir5ay) ([raw file](https://github.com/cockroachdb/cockroach/blob/7797919d7db3b82bb816d12f18ce3e6596a429e8/pkg/sql/alter_index.go#L94)):*
> ```Go
> 			}
> 		default:
> 			return fmt.Errorf("unsupported alter cmd: %T", cmd)
> ```

We should probably spell out "command", but I guess it's like this in ALTER TABLE, so ¯\_(ツ)_/¯  

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20957#-:-L0u_m-C0FSnoVw7a-JW:b-1qphvw)*

**Comment by danhhz** [comment]
Agreed, it's worth getting an extra set of eyes. I'll poke around

---

Review status: 1 of 16 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[pkg/sql/alter_index.go, line 70 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20957#-L0u_N_nrsHy36RNcne8:-L1OGfZMfVEdKTc5G4Q9:b-7iqpnr) ([raw file](https://github.com/cockroachdb/cockroach/blob/7797919d7db3b82bb816d12f18ce3e6596a429e8/pkg/sql/alter_index.go#L70)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

No longer needed right?
</blockquote></details>

I know it's nitpicky, but I'd prefer to wait until the rfc update merges before we start ripping all this out

---

*[pkg/sql/alter_index.go, line 82 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20957#-L0u_eZbyFqAjJkGWS-o:-L1OGldwrnEylXh9GE_a:b-89a4l7) ([raw file](https://github.com/cockroachdb/cockroach/blob/7797919d7db3b82bb816d12f18ce3e6596a429e8/pkg/sql/alter_index.go#L82)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Can drop this, right?
</blockquote></details>

Ditto

---

*[pkg/sql/alter_index.go, line 94 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20957#-L0u_SvwxkjFUNtAu-S2:-L1OH0o7b0um-g0ZRmpW:b-4s6idd) ([raw file](https://github.com/cockroachdb/cockroach/blob/7797919d7db3b82bb816d12f18ce3e6596a429e8/pkg/sql/alter_index.go#L94)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

We should probably spell out "command", but I guess it's like this in ALTER TABLE, so ¯\_(ツ)_/¯  
</blockquote></details>

not anymore : - )

---

*[pkg/sql/alter_index.go, line 125 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20957#-L0prdkq1cwmkbAg7W3D-r1-125:-L1OHE5wNwYm6AwimrcX:b-sc86nu) ([raw file](https://github.com/cockroachdb/cockroach/blob/7797919d7db3b82bb816d12f18ce3e6596a429e8/pkg/sql/alter_index.go#L125)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

Intentionally leaving this TODO undone until I make sure that no-one is going to have strong objections to adding this generic CreateIndex
</blockquote></details>

okay did this

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20957#-:-L1OI26ksW5fnQGXea94:b78fw6s)*

**Comment by danhhz** [comment]
rfc change merged, so I removed the TODOs

**Comment by danhhz** [comment]
Hey @jordanlewis or @knz, any thoughts on the AlterIndex I added to parallel AlterTable?

**Comment by jordanlewis** [comment]
SQL changes :lgtm:! You're missing logic tests, but I assume you guys aren't writing those for partitioning, preferring to live in partition_test.go?

---

Reviewed 1 of 12 files at r1, 15 of 15 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20957#-:-L1TZ5SEvrtgyx56_UIR:b-865lnh)*

**Comment by danhhz** [comment]
TFTRs! Yeah, in general, it's easier to test ccl code in the ccl tree. The partition_test.go stuff should cover the interesting cases

**Comment by jordanlewis** [comment]
Mark added CCL logic tests!

**Comment by danhhz** [comment]
I saw! I guess the real issue is I'm not sure what cases they'd cover that aren't already covered in partition_test.go

**Pull request merged**